Adopt dynamicDowncast<> in ReplaceSelectionCommand

annevk opened the pull request:
#### f7b69f5ea804100eb22e0a312c074a59425d4453
<pre>
Adopt dynamicDowncast&lt;&gt; in ReplaceSelectionCommand
<a href="https://bugs.webkit.org/show_bug.cgi?id=267701">https://bugs.webkit.org/show_bug.cgi?id=267701</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::isInterchangeNewlineNode):
(WebCore::isInterchangeConvertedSpaceSpan):
(WebCore::ReplacementFragment::removeInterchangeNodes):
(WebCore::haveSameTagName):
(WebCore::ReplaceSelectionCommand::shouldMerge):
(WebCore::fragmentNeedsColorTransformed):
(WebCore::ReplaceSelectionCommand::inverseTransformColor):
(WebCore::ReplaceSelectionCommand::removeRedundantStylesAndKeepStyleSpanInline):
(WebCore::ReplaceSelectionCommand::makeInsertedContentRoundTrippableWithHTMLTreeBuilder):
(WebCore::ReplaceSelectionCommand::moveNodeOutOfAncestor):
(WebCore::ReplaceSelectionCommand::removeUnrenderedTextNodesAtEnds):
(WebCore::isInlineNodeWithStyle):
(WebCore::ReplaceSelectionCommand::shouldPerformSmartReplace const):
(WebCore::ReplaceSelectionCommand::addSpacesForSmartReplace):
(WebCore::ReplaceSelectionCommand::mergeTextNodesAroundPosition):
(WebCore::ReplaceSelectionCommand::insertAsListItems):
(WebCore::ReplaceSelectionCommand::performTrivialReplace):

Canonical link: <a href="https://commits.webkit.org/273354@main">https://commits.webkit.org/273354@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/607d3b4501f85f2971817df6dbf24faf6bd0a4a2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35032 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37118 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37786 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31645 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36191 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16315 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11009 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30581 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35575 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11806 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31248 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10349 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10409 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31341 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39036 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31870 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31670 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36427 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10512 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8443 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34515 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12309 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8055 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11044 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11374 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/editing/ReplaceSelectionCommand.cpp:
```diff
@@ -550,7 +556,7 @@ bool ReplaceSelectionCommand::shouldMerge(const VisiblePosition& source, const V
         && enclosingListChild(sourceBlock.get()) == enclosingListChild(destinationNode.get())
         && enclosingTableCell(source.deepEquivalent()) == enclosingTableCell(destination.deepEquivalent())
-        && (!isHeaderElement(*sourceBlock) || haveSameTagName(sourceBlock.get(), destinationBlock.get()))
+        && (!isHeaderElement(*sourceBlock) || haveSameTagName(*sourceBlock, *destinationBlock))
```
destinationBlock cannot be assumed to be non-null.

rniwa commented on Source/WebCore/editing/ReplaceSelectionCommand.cpp:
```diff
@@ -531,9 +533,13 @@ static bool isHeaderElement(const Node& a)
 }
 
-static bool haveSameTagName(Node* a, Node* b)
+static bool haveSameTagName(Node& a, RefPtr<Node> b)
```
It strikes me as strange that one of them is a reference and the other one is `RefPtr`.

annevk replied:
I could have added a null check to the caller. I suggest we revisit this once we locate more potential callers for this. Some light grepping doesn't reveal any obvious pattern and it seems a bit suspect to compare the tag name (rather than local name) to begin with.

cdumez replied:
@annevk Could you please revert this change? It is not OK to pass a RefPtr<> like this as parameter. This causes ref-counting churn.

annevk replied:
In case anyone reads this in the future, #23106 fixes this.

annevk committed:
Adopt dynamicDowncast<> in ReplaceSelectionCommand
https://bugs.webkit.org/show_bug.cgi?id=267701

Reviewed by Ryosuke Niwa.

* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::isInterchangeNewlineNode):
(WebCore::isInterchangeConvertedSpaceSpan):
(WebCore::ReplacementFragment::removeInterchangeNodes):
(WebCore::haveSameTagName):
(WebCore::ReplaceSelectionCommand::shouldMerge):
(WebCore::fragmentNeedsColorTransformed):
(WebCore::ReplaceSelectionCommand::inverseTransformColor):
(WebCore::ReplaceSelectionCommand::removeRedundantStylesAndKeepStyleSpanInline):
(WebCore::ReplaceSelectionCommand::makeInsertedContentRoundTrippableWithHTMLTreeBuilder):
(WebCore::ReplaceSelectionCommand::moveNodeOutOfAncestor):
(WebCore::ReplaceSelectionCommand::removeUnrenderedTextNodesAtEnds):
(WebCore::isInlineNodeWithStyle):
(WebCore::ReplaceSelectionCommand::shouldPerformSmartReplace const):
(WebCore::ReplaceSelectionCommand::addSpacesForSmartReplace):
(WebCore::ReplaceSelectionCommand::mergeTextNodesAroundPosition):
(WebCore::ReplaceSelectionCommand::insertAsListItems):
(WebCore::ReplaceSelectionCommand::performTrivialReplace):

Canonical link: https://commits.webkit.org/273354@main
